File: nodejs/examples/extraction/03_schema_generation_http.ts
#!/usr/bin/env npx ts-node
/**
 * Schema Generation with HTTP
 *
 * This example shows how to automatically generate CSS extraction schemas
 * from HTML using LLM via direct HTTP API calls (no SDK).
 *
 * Usage:
 *     npx ts-node 03_schema_generation_http.ts
 *
 * Requirements:
 *     Node.js 18+ (for native fetch)
 */

// Configuration
const API_URL = 'https://api.crawl4ai.com';
const API_KEY = 'YOUR_API_KEY'; // Replace with your API key

async function generateExtractionSchema(): Promise<void> {
  const headers = {
    'X-API-Key': API_KEY,
    'Content-Type': 'application/json',
  };

  // First, get the HTML content
  console.log('Fetching Hacker News HTML...');
  const crawlResponse = await fetch(`${API_URL}/v1/crawl`, {
    method: 'POST',
    headers,
    body: JSON.stringify({
      url: 'https://news.ycombinator.com',
      strategy: 'http',
    }),
  });

  if (!crawlResponse.ok) {
    console.log(`Error: ${crawlResponse.status} - ${await crawlResponse.text()}`);
    return;
  }

  const crawlData = await crawlResponse.json();
  const html = crawlData.html || '';
  console.log(`Got ${html.length} bytes of HTML`);

  // Generate schema using LLM
  console.log('\nGenerating CSS extraction schema...');
  const schemaResponse = await fetch(`${API_URL}/v1/tools/schema`, {
    method: 'POST',
    headers,
    body: JSON.stringify({
      html,
      query: 'Extract all stories with their title, URL, points, and author',
      schema_type: 'CSS',
    }),
  });

  if (!schemaResponse.ok) {
    console.log(`Error: ${schemaResponse.status} - ${await schemaResponse.text()}`);
    return;
  }

  const schemaData = await schemaResponse.json();

  if (schemaData.error) {
    console.log(`Error: ${schemaData.error}`);
    return;
  }

  const schema = schemaData.schema;
  console.log('\nGenerated Schema:');
  console.log(JSON.stringify(schema, null, 2));

  // Now use the generated schema for extraction
  console.log('\n\nTesting generated schema...');
  const extractResponse = await fetch(`${API_URL}/v1/crawl`, {
    method: 'POST',
    headers,
    body: JSON.stringify({
      url: 'https://news.ycombinator.com',
      strategy: 'http',
      crawler_config: {
        extraction_strategy: {
          type: 'json_css',
          schema,
        },
      },
    }),
  });

  if (!extractResponse.ok) {
    console.log(`Error: ${extractResponse.status} - ${await extractResponse.text()}`);
    return;
  }

  const extractData = await extractResponse.json();
  const stories = extractData.extracted_content || [];
  console.log(`\nExtracted ${stories.length} stories`);
  console.log('\nFirst 2 stories:');
  stories.slice(0, 2).forEach((story: any) => {
    console.log(`\n${JSON.stringify(story, null, 2)}`);
  });
}

generateExtractionSchema().catch(console.error);

File: python/examples/extraction/03_schema_generation_http.py
#!/usr/bin/env python3
"""
Schema Generation with HTTP

This example shows how to automatically generate CSS extraction schemas
from HTML using LLM via direct HTTP API calls (no SDK).

Usage:
    python 03_schema_generation_http.py

Requirements:
    pip install httpx
"""

import httpx

# Configuration
API_URL = "https://api.crawl4ai.com"
API_KEY = "your_api_key_here"  # Replace with your API key

def generate_extraction_schema():
    """Generate CSS schema for Hacker News stories via HTTP API."""

    headers = {
        "X-API-Key": API_KEY,
        "Content-Type": "application/json"
    }

    # First, get the HTML content
    print("Fetching Hacker News HTML...")
    crawl_response = httpx.post(
        f"{API_URL}/v1/crawl",
        headers=headers,
        json={
            "url": "https://news.ycombinator.com",
            "strategy": "http"
        },
        timeout=60.0
    )

    if crawl_response.status_code != 200:
        print(f"Error: {crawl_response.status_code} - {crawl_response.text}")
        return

    html = crawl_response.json().get("html", "")
    print(f"Got {len(html)} bytes of HTML")

    # Generate schema using LLM
    print("\nGenerating CSS extraction schema...")
    schema_response = httpx.post(
        f"{API_URL}/v1/tools/schema",
        headers=headers,
        json={
            "html": html,
            "query": "Extract all stories with their title, URL, points, and author",
            "schema_type": "CSS"
        },
        timeout=60.0
    )

    if schema_response.status_code != 200:
        print(f"Error: {schema_response.status_code} - {schema_response.text}")
        return

    schema_data = schema_response.json()

    if schema_data.get("error"):
        print(f"Error: {schema_data['error']}")
        return

    schema = schema_data["schema"]
    print("\nGenerated Schema:")
    print(schema)

    # Now use the generated schema for extraction
    print("\n\nTesting generated schema...")
    extract_response = httpx.post(
        f"{API_URL}/v1/crawl",
        headers=headers,
        json={
            "url": "https://news.ycombinator.com",
            "strategy": "http",
            "crawler_config": {
                "extraction_strategy": {
                    "type": "json_css",
                    "schema": schema
                }
            }
        },
        timeout=60.0
    )

    if extract_response.status_code != 200:
        print(f"Error: {extract_response.status_code} - {extract_response.text}")
        return

    stories = extract_response.json().get("extracted_content", [])
    print(f"\nExtracted {len(stories)} stories")
    print("\nFirst 2 stories:")
    for story in stories[:2]:
        print(f"\n{story}")

if __name__ == "__main__":
    generate_extraction_schema()
